### PR TITLE
Expose counterparty forwarding info in ChannelDetails

### PR DESCRIPTION
**Pull request by valentinewallace**
Supersedes #207. Tested in the sample :)

### CODE REVIEW

File: lightning/src/ln/channel.rs [hunk 1]
```diff
@@ -4114,6 +4127,26 @@ impl<Signer: Sign> Channel<Signer> {
...
+	pub fn get_counterparty_forwarding_info(&self) -> Option<CounterpartyForwardingInfo> {
+		self.counterparty_forwarding_info.clone()
+	}
+
+	pub fn channel_update(&mut self, msg: &msgs::ChannelUpdate) -> Result<(), ChannelError> {
+		let usable_channel_value_msat = (self.channel_value_satoshis - self.counterparty_selected_channel_reserve_satoshis) * 1000;
+		if msg.contents.htlc_minimum_msat >  usable_channel_value_msat {
+			return Err(ChannelError::Close("Minimum htlc value is greater than channel value".to_string()));
+		}
+		self.counterparty_htlc_minimum_msat = msg.contents.htlc_minimum_msat;
```
Comment (TheBlueMatt) [inline lightning/src/ln/channel.rs:4140]: Isn't this *our* htlc_minimum? Its the value that the counterparty is announcing to the world as "the minimum HTLC value (in millisatoshi) that the channel peer will accept.", and I think *we're" the "channel peer".

Comment (valentinewallace) [reply lightning/src/ln/channel.rs]: Oops, my bad. Hm, so the counterparty is unable to update the min htlc that they'll accept, then? Weird.

File: lightning/src/ln/channel.rs [hunk 2]
```diff
@@ -4114,6 +4127,26 @@ impl<Signer: Sign> Channel<Signer> {
...
+	pub fn get_counterparty_forwarding_info(&self) -> Option<CounterpartyForwardingInfo> {
+		self.counterparty_forwarding_info.clone()
+	}
+
+	pub fn channel_update(&mut self, msg: &msgs::ChannelUpdate) -> Result<(), ChannelError> {
+		let usable_channel_value_msat = (self.channel_value_satoshis - self.counterparty_selected_channel_reserve_satoshis) * 1000;
+		if msg.contents.htlc_minimum_msat >  usable_channel_value_msat {
+			return Err(ChannelError::Close("Minimum htlc value is greater than channel value".to_string()));
+		}
+		self.holder_htlc_minimum_msat = msg.contents.htlc_minimum_msat;
```
Comment (TheBlueMatt) [inline lightning/src/ln/channel.rs:4140]: We shouldnt update this - we set our own htlc_minimum, if the counterparty sets a higher one when relaying to us, that's fine, but they can also decrease it later. Either way, they must not violate our original htlc_minimum.

File: lightning/src/ln/channel.rs [hunk 3]
```diff
@@ -281,6 +281,13 @@ impl HTLCCandidate {
 	}
 }
 
+#[derive(Clone)]
+pub struct CounterpartyForwardingInfo {
```
Comment (TheBlueMatt) [inline lightning/src/ln/channel.rs:285]: Hmm, so this isn't *actually* public because `channel` is a `pub(crate)` module (hence why its not a compile failure that there are no docs).

Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs:285]: I guess I'm confused how this worked on the sample? Can you access the fields?

Comment (valentinewallace) [reply lightning/src/ln/channel.rs:285]: Oh that's weird. Yeah, I can access the fields in the sample.

Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs:285]: Yea, I think we should either re-export it or just move it to `ChannelManager`, right now you can't create one, which maybe isnt that important but seems strange. Moving it would also enforce our compile-time checks for missing docs, which would be good :)

File: lightning/src/ln/channel.rs [hunk 4]
```diff
@@ -4114,6 +4127,25 @@ impl<Signer: Sign> Channel<Signer> {
...
 
+	/// Get forwarding information for the counterparty.
+	pub fn get_counterparty_forwarding_info(&self) -> Option<CounterpartyForwardingInfo> {
+		self.counterparty_forwarding_info.clone()
+	}
+
+	pub fn channel_update(&mut self, msg: &msgs::ChannelUpdate) -> Result<(), ChannelError> {
+		let usable_channel_value_msat = (self.channel_value_satoshis - self.counterparty_selected_channel_reserve_satoshis) * 1000;
+		if msg.contents.htlc_minimum_msat >  usable_channel_value_msat {
+			return Err(ChannelError::Close("Minimum htlc value is greater than channel value".to_string()));
```
Comment (ariard) [inline lightning/src/ln/channel.rs:4138]: I think this check is purposeful to make sure when we initiate channel with respect our own config or that other counterparty don't waste time/fees with an unroutable channel, but better to move it in `new_from_req`/`new_outbound`, not already present ?

File: lightning/src/ln/channel.rs (hunk 3 above)
Comment (ariard) [inline lightning/src/ln/channel.rs:285]: I had a look on #207, effectively this make sense to learn counterparty forwarding policy on this link to communicate a routing hint to a payee. I think this struct should document this usage, as otherwise it's a bit confusing why we need to observe peer's `channel_update` on a local channel.

Should we also add `htlc_maximum_msat` ? Don't issue an invoice with an amount above max otherwise your hint is useless ?

**Review by ariard — COMMENTED** [review]
Thanks to resuscitate this, almost good to me

File: lightning/src/ln/channel.rs (hunk 3 above)
Comment (valentinewallace) [reply lightning/src/ln/channel.rs:286]: > struct should document this usage, as otherwise it's a bit confusing why we need to observe peer's channel_update on a local channel.

Documented!

> Should we also add htlc_maximum_msat ? Don't issue an invoice with an amount above max otherwise your hint is useless ?

Hm, but with MPP, the total invoice amount may violate an individual channel's `htlc_max_msat` but not the single payment path that uses said channel. :thinking:

File: lightning/src/ln/channel.rs (hunk 4 above)
Comment (valentinewallace) [reply lightning/src/ln/channel.rs:4142]: The check is present in `new_from_req` (`new_outbound` isn't parameterized by `htlc_min_msat`). Why wouldn't we also check here, though?

File: lightning/src/ln/channel.rs [hunk 5]
```diff
@@ -281,6 +281,17 @@ impl HTLCCandidate {
...
 }
 
+/// Information needed for constructing an invoice route hint for this channel.
+#[derive(Clone)]
+pub struct CounterpartyForwardingInfo {
+	/// Base routing fee in millisatoshis.
+	pub fee_base_msat: u32,
+	/// Amount in millionths of a satoshi the channel will charge per transferred satoshi.
+	pub fee_proportional_millionths: u32,
+	/// The number of blocks to subtract from incoming HTLC cltv_expiry values.
```
Comment (TheBlueMatt) [inline lightning/src/ln/channel.rs:291]: nit: if other comments come up, can you reword this as its pretty unclear to me? The wording for the similar field in UnsignedChannelUpdate is `/// The number of blocks to subtract from incoming HTLC cltv_expiry values` though that could probably also use a "to calculate the outbound HTLC cltv_expiry value" and a "minimum" somewhere in it.

File: lightning/src/ln/channelmanager.rs [hunk 6]
```diff
@@ -574,6 +577,9 @@ pub struct ChannelDetails {
 	/// True if the channel is (a) confirmed and funding_locked messages have been exchanged, (b)
 	/// the peer is connected, and (c) no monitor update failure is pending resolution.
 	pub is_live: bool,
+
+	/// Counterparty forwarding information.
```
Comment (TheBlueMatt) [inline lightning/src/ln/channelmanager.rs:581]: nit: could maybe be expanded to say, like, "information on the fees and requirements that the counterparty requires when forwarding payments to us through this channel".

**Review by TheBlueMatt — APPROVED** [review]
Looks good, a few doc nits, but dont bother unless you think you should or you're addressing other comments.

File: lightning/src/ln/peer_handler.rs [hunk 7]
```diff
@@ -142,6 +142,8 @@ impl ChannelMessageHandler for ErroringMessageHandler {
 	fn handle_channel_reestablish(&self, their_node_id: &PublicKey, msg: &msgs::ChannelReestablish) {
 		ErroringMessageHandler::push_error(self, their_node_id, msg.channel_id);
 	}
+	// msgs::ChannelUpdate do not contain the channel_id field, so we just drop them.
```
Comment (jkczyz) [inline lightning/src/ln/peer_handler.rs:145]: nit: s/do/does

File: lightning/src/ln/peer_handler.rs [hunk 8]
```diff
@@ -970,6 +972,7 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref> PeerManager<D
 				}
 			},
 			wire::Message::ChannelUpdate(msg) => {
+				self.message_handler.chan_handler.handle_channel_update(&peer.their_node_id.unwrap(), &msg);
 				let should_forward = match self.message_handler.route_handler.handle_channel_update(&msg) {
```
Comment (jkczyz) [inline lightning/src/ln/peer_handler.rs:976]: If checks in `route_handler` fail, should we still send this message to `chan_handler`?

File: lightning/src/ln/channel.rs [hunk 9]
```diff
@@ -4130,6 +4131,11 @@ impl<Signer: Sign> Channel<Signer> {
 		}
 	}
 
+	/// Get forwarding information for the counterparty.
+	pub fn get_counterparty_forwarding_info(&self) -> Option<CounterpartyForwardingInfo> {
```
Comment (jkczyz) [inline lightning/src/ln/channel.rs:4135]: Not sure if we've been entirely consistent here, but consider dropping the `get_` prefix for accessor methods since that tends to be the Rust convention.

File: lightning/src/ln/channel.rs [hunk 10]
```diff
@@ -5021,6 +5028,55 @@ mod tests {
...
+		assert_eq!(node_a_chan.holder_htlc_minimum_msat, 1);
+		assert!(node_a_chan.get_counterparty_forwarding_info().is_some());
+		match node_a_chan.counterparty_forwarding_info {
+			Some(info) => {
+				assert_eq!(info.cltv_expiry_delta, 100);
+				assert_eq!(info.fee_base_msat, 110);
+				assert_eq!(info.fee_proportional_millionths, 11);
+			},
+			None => panic!("expected counterparty forwarding info to be Some")
+		}
```
Comment (jkczyz) [inline lightning/src/ln/channel.rs:5077]: Can these two assertions be combined?

**Review by jkczyz — COMMENTED** [review]
Mostly nits but wondering if failures handling in the router should dictate if the channel manager should still handle the message. See comment.

File: lightning/src/ln/peer_handler.rs (hunk 8 above)
Comment (TheBlueMatt) [reply lightning/src/ln/peer_handler.rs:976]: We'd need some way to differentiate "we haven't seen a channel_announcement/the channel is private" as a failure from "invalid signature" or other failures. We /could/ check the signature in ChannelManager, but we know it came directly from our peer, so not sure if we need to.

Comment (jkczyz) [reply lightning/src/ln/peer_handler.rs:976]: I see what you are saying for this particular message. Though I do wonder more broadly if we are expecting custom handler implementations to re-implement all BOLT checks or if there is some subset we should check regardless of implementation. Definitely outside the scope of this PR.

Comment (TheBlueMatt) [reply lightning/src/ln/peer_handler.rs:976]: Yea, thats a good question. I kinda figure there's relatively limited uses for implementing a full custom message handler, and its more about the message handlers being optional/dummy in cases where, eg, you don't want to have a network graph so I was never too worried about it.

File: lightning/src/ln/channel.rs (hunk 5 above)
Comment (valentinewallace) [reply lightning/src/ln/channel.rs]: I stole the wording word-for-word from the wording you quoted^, lol. But, updated the comment in both places.

File: lightning/src/ln/peer_handler.rs (hunk 8 above)
Comment (jkczyz) [reply lightning/src/ln/peer_handler.rs:976]: I suppose the wrapper approach in a similar vein as `Read` that I had mentioned [elsewhere](https://github.com/rust-bitcoin/rust-lightning/pull/814#discussion_r588846888) may be sufficient.

File: lightning/src/ln/channel.rs (hunk 5 above)
Comment (jkczyz) [reply lightning/src/ln/channel.rs]: I'm more of the mind of having a summary comment in one place and more detailed when where warranted. Otherwise, the documentation is at risk of going out of sync.

Comment (valentinewallace) [reply lightning/src/ln/channel.rs]: Adapted to @jkczyz's suggestion!

**Review by jkczyz — APPROVED** [review]

File: lightning/src/ln/msgs.rs [hunk 11]
```diff
@@ -543,7 +543,12 @@ pub struct UnsignedChannelUpdate {
 	pub timestamp: u32,
 	/// Channel flags
 	pub flags: u8,
-	/// The number of blocks to subtract from incoming HTLC cltv_expiry values
+	/// The number of blocks such that if:
+	/// `htlc.cltv_expiry - current_block_height <= cltv_expiry_delta`
+	/// then we need to fail the HTLC backwards. When forwarding an HTLC, cltv_expiry_delta is used to
+	/// calculate the outgoing HTLC's cltv_expiry value -- so, if an incoming HTLC comes in with a
```
Comment (TheBlueMatt) [inline lightning/src/ln/msgs.rs:549]: This is a big confusing, we never calculate cltv_expiry value, its specified for us in the route, we just require that the difference between the inbound and outbound legs be at least X.

Comment (TheBlueMatt) [reply lightning/src/ln/msgs.rs:549]: Same goes for the docs in CounterpartyForwardingInfo::cltv_expiry_delta - we don't add the value, we (well, really our counterparty) just use it as a check.

Comment (valentinewallace) [reply lightning/src/ln/msgs.rs:549]: Ah, that's a good distinction. Let me know if the updated comment looks good and I'll squash it in.

Comment (TheBlueMatt) [reply lightning/src/ln/msgs.rs:549]: I think we should avoid the word "calculate in the second sentence. Also, the check that's referenced isn't the relevant one, its the `if (msg.cltv_expiry as u64) < (*outgoing_cltv_value) as u64 + CLTV_EXPIRY_DELTA as u64` one.

Comment (valentinewallace) [reply lightning/src/ln/msgs.rs]: Updated!

Comment (TheBlueMatt) [reply lightning/src/ln/msgs.rs]: Looks good to me!

**Comment by valentinewallace** [comment]
CI passed ;)

**Review by TheBlueMatt — APPROVED** [review]
Only changes since Jeff's ACK were two small doc fixes.
